Feat: VTEX API Schema extensions (#1903)
vtex/faststore: apps/site/pages/docs/api-extensions.mdx

---
title: '  Overview'
---

import { Steps, Step } from 'nextra-theme-docs'

<header className="hero">

# API Extensions

> In this guide you will learn how to extends the FastStore API schema.

Reviewer: ```suggestion
In this guide you will learn how to extend the FastStore API schema.
```

</header>

> Although the [FastStore API](https://v1.faststore.dev/reference/api/faststore-api) provides a complete [GraphQL schema for ecommerce](https://v1.faststore.dev/reference/api/queries), some stores may need to access other, more specific, information.

Reviewer: ```suggestion
Although the [FastStore API](https://v1.faststore.dev/reference/api/faststore-api) provides a [GraphQL schema for ecommerce](https://v1.faststore.dev/reference/api/queries) that suits a general use case, some stores may need to access other, more specific, information.
```

> You should not add requests to other APIs to your storefront since this may [impact site performance](https://v1.faststore.dev/how-to-guides/faststore-api/fetching-api-data#best-practices-for-fetching-data). For those cases, it is possible to extend the FastStore API schema, adding new data to the existing [queries](https://v1.faststore.dev/reference/api/queries).

Reviewer: ```suggestion
You will not be able to add other APIs endpoints to your storefront since this may [impact site performance](https://v1.faststore.dev/how-to-guides/faststore-api/fetching-api-data#best-practices-for-fetching-data). If you need to fetch more data, it is possible to extend the FastStore API schema, adding new data to the existing [queries](https://v1.faststore.dev/reference/api/queries).
```

---

## Before starting

Note that even though you can add information to the FastStore API schema, you must be careful not to over-fetch data on your pages. See the [best practices for fetching data on your storefront](https://v1.faststore.dev/how-to-guides/faststore-api/fetching-api-data#best-practices-for-fetching-data).

> As stores tend to grow, so does the possibility of using data from VTEX APIs that the FastStore API does not expose.

Reviewer: ```suggestion
As stores grow, so does the possibility of using data from VTEX APIs that the FastStore API does not expose.
```

Another possibility is to use data that is provided by third-party APIs. In both cases, FastStore provides these features by using API Extensions.

Inside the repository, there is a clear separation about what comes from VTEX APIs and what comes from third-party APIs. First, let's take a look in the VTEX APIs Schema extension possibility.

---

## VTEX APIs Schema Extensions

FastStore streamline the way of using data from VTEX APIs that the FastStore API does not expose.

> As FastStore uses GraphQL, it's important to define its typeDefs and resolvers when desires creates APIs Extensions by following the steps below

Reviewer: ```suggestion
As FastStore uses GraphQL, it's required to write type definitions and resolvers to fetch the data you need.  You will learn how to do this using API Extensions by following the steps below.
```

Maybe it's worthwhile to add a link to type definition and resolvers for GraphQL.

Author: Nice! Thank you. I'll add the 
https://graphql.org/learn/schema/#object-types-and-fields
and https://graphql.org/learn/execution/#root-fields-resolvers
links.

### Folder Structure

> The store's repository provides the following folder structure for the VTEX APIs Schema Extensions that is hold in the `src>graphql>vtex` folder.

Reviewer: ```suggestion
The store's repository provides the following folder structure for the VTEX APIs Schema Extensions that is held in the `src/graphql/vtex` folder.
```

```ts
starter.store/
└─ src/
    └─ graphql/
        └─ vtex/
          ├─ resolvers/
              └─ index.ts
              └─ <resolverName>.ts
              └─ …
          ├─ typeDefs/
              └─<typeName>.graphql
              └─ …
```

### TypeDefs

Your new type definitions set the data structure for your new fields, extended from the existing FastStore API GraphQL queries and types.

> You can create new `<typeName>.graphql` files inside the `vtex>typeDefs` folder. The way that it will be organized inside this folder is flexible, it's possible to create multiple typeDefs files or group them all together in just one file.

Reviewer: ```suggestion
You can create a new `<typeName>.graphql` files inside the `vtex/typeDefs` folder. The way that it will be organized inside this folder is flexible, it's possible to create multiple typeDefs files or group them all together in just one file.
```
